src/utils/viewer/viewer.cpp

```diff
@@ -359,7 +368,46 @@ void Viewer::drawEvent() {
   int DEFAULT_SCENE = 0;
   int sceneID = sceneID_[DEFAULT_SCENE];
   auto& sceneGraph = sceneManager_.getSceneGraph(sceneID);
-  renderCamera_->getMagnumCamera().draw(sceneGraph.getDrawables());
+
+  // XXX
+  auto& drawables = sceneGraph.getDrawables();
+  auto total = drawables.size();
+  size_t visibles = 0;
+  auto& cam = renderCamera_->getMagnumCamera();
+  if (enableFrustumCulling_) {
+    Magnum::SceneGraph::DrawableGroup3D visibleDrawables;
+
+    for (int iDrawable = total - 1; iDrawable >= 0; iDrawable--) {
+      // This, of course, will cause segfault.
+      // for (int iDrawable = 0; iDrawable < total; ++iDrawable)
+      scene::SceneNode& node =
+          dynamic_cast<scene::SceneNode&>(drawables[iDrawable].object());
+      // mvp = proj * modelview
+      auto mvp = cam.projectionMatrix() * cam.cameraMatrix() *
+                 node.absoluteTransformation();
+      Magnum::Math::Frustum<float> frustum;
+      frustum = Magnum::Math::Frustum<float>::fromMatrix(mvp);
+
+      if (Magnum::Math::Intersection::rangeFrustum(node.getMeshBB(), frustum)) {
```

Comment:
Since >99% of the scene is currently static, I think it would make sense to have the AABBs *absolute*, to avoid the need to do all the above matrix ops again per each node -- basically calculating the frustum just from the current camera position/orientation.

And for dynamic/physics objects, to avoid having to recalculate the AABBs after each physics tick (unless Bullet can give you absolute AABBs "for free"), what about having bounding spheres instead? Then you would only need absolute translation of each and could again use a single `frustum` for all nodes. It *may* result in slightly more stuff being rendered than needed, but I think that would be balanced by simpler handling here.